src/net.cpp

```diff
@@ -834,6 +846,37 @@ CNetMessage V2TransportDeserializer::GetMessage(const CMessageHeader::MessageSta
     msg.m_message_size = msg.m_recv.size();                                                                    //message payload size (excluding command)
     msg.m_raw_message_size = CHACHA20_POLY1305_AEAD_AAD_LEN + m_message_size + CHACHA20_POLY1305_AEAD_TAG_LEN; // raw wire size
 
+    const int64_t now = GetTime();
+    if (m_rekey_flag) {
+        if (!gArgs.GetBoolArg("-netencryptionfastrekey", false) && (m_time_last_rekey + MIN_REKEY_TIME > now)) {
+            // remote peer was not respecting the mininal rekey time (DoS)
+            LogPrint(BCLog::NET, "Invalid rekey (DoS)\n");
+            msg.m_valid_checksum = false;
+            msg.m_valid_header = false;
+        } else {
+            // make sure we rekey at this point, next message is supposed to be encrypted with the new key
+            CHash256().Write(m_session_id.begin(), m_session_id.size()).Write(m_aead_k1.data(), m_aead_k1.size()).Finalize(m_aead_k1.data());
+            CHash256().Write(m_session_id.begin(), m_session_id.size()).Write(m_aead_k2.data(), m_aead_k2.size()).Finalize(m_aead_k2.data());
+
+            // reset the AEAD context
+            m_aead.reset(new ChaCha20Poly1305AEAD(m_aead_k1.data(), m_aead_k1.size(), m_aead_k2.data(), m_aead_k2.size()));
+            LogPrint(BCLog::NET, "Rekey: new recv keys (%s, %s)\n", HexStr(m_aead_k1), HexStr(m_aead_k2));
+
+            // reset sequence numbers
+            m_payload_seqnr = 0;
+            m_aad_seqnr = 0;
+            m_aad_pos = 0;
+            m_bytes_decrypted = 0;
+            m_time_last_rekey = now;
+        }
+    } else if (m_bytes_decrypted > REKEY_ABORT_LIMIT_BYTES || now - m_time_last_rekey > REKEY_ABORT_LIMIT_TIME ||
+               (gArgs.GetBoolArg("-netencryptionfastrekey", false) && m_bytes_decrypted > 64 * 1024)) {
+        // don't further decrypt and therefore abort connection when counterparty failed to respect rekey limits
```

Comment:
Where do we abort connection based on invalid checksum and header? It's only checked in `ProcessMessages` but doesn't trigger a connection AFAICT?

Reply:
Yes. We don't right now as this is currently identical to V1 (where we don't disconnect if a message has an invalid header or an invalid SHA256 checksum).
See #15206 and #15197.

Reply:
Thanks, will review them with context.